### PR TITLE
Add functions for verbose and with context execution

### PR DESCRIPTION
**Pull request by farshidtz**
Adding additional functions for verbose execution, as well as support for passing context (for early interruption).

### CODE REVIEW

**Comment by farshidtz** [comment]
@locnnil please feel free to add more comments to the same branch for the remaining refactoring and test implementations.

**Comment by locnnil** [comment]
@farshidtz Is there some specific reason to [run the tests 100 times](https://github.com/canonical/matter-snap-testing/blob/2f282dddac850ee973c84f055c277be01592b684/.github/workflows/unit-testing.yml#L24)?

**Comment by farshidtz** [comment]
> @farshidtz Is there some specific reason to [run the tests 100 times](https://github.com/canonical/matter-snap-testing/blob/2f282dddac850ee973c84f055c277be01592b684/.github/workflows/unit-testing.yml#L24)?

I remember we had a race condition which was causing a failure around 10-20% of the time. The Go's data race detector wasn't catching it either. I can't recall the specifics, but it should be recorded somewhere. We fixed the bug and continued running the tests that way.

File: utils/exec_test.go [hunk 1]
```diff
@@ -30,24 +31,53 @@ func TestExec(t *testing.T) {
...
+	t.Run("context with timeout", func(t *testing.T) {
+		ctx, cancel := context.WithDeadline(context.Background(), <-time.After(2*time.Second))
+		defer cancel()
+
+		stdout, stderr, err := exec(nil, ctx, `sleep infinity`, true)
+		assert.Error(t, err)
+		t.Logf("stdout: %s", stdout)
+		t.Logf("stdeer: %s", stderr)
+	})
 }
```
Comment (farshidtz) [inline utils/exec_test.go:83]: Subtest titles don't match what is being tested.

File: utils/exec_test.go [hunk 2]
```diff
@@ -30,24 +31,53 @@ func TestExec(t *testing.T) {
...
+		stdout, stderr, err := exec(t, nil, `echo "failing" >&2; echo "succeeding"`, true)
 		assert.NoError(t, err)
 		assert.Equal(t, "failing\n", stderr)
 		assert.Equal(t, "succeeding\n", stdout)
 	})
 
+	t.Run("context with cancel", func(t *testing.T) {
+		timeout := 5 * time.Second
+		ctx, cancel := context.WithTimeout(context.Background(), timeout)
+		defer cancel()
```
Comment (farshidtz) [inline utils/exec_test.go:57]: Defer should not be used in tests that can fail and exit without returning. In such cases, `cancel()` will not be called. See instead https://pkg.go.dev/testing#T.Cleanup

File: utils/exec_test.go [hunk 3]
```diff
@@ -30,24 +31,53 @@ func TestExec(t *testing.T) {
...
+			t.Errorf("Unexpected error: %v", err)
+		}
+
+		if stdout != "" {
+			t.Fatalf("Unexpected stdout: %s", stdout)
+		}
+
+		if stderr != "" {
+			t.Fatalf("Unexpected stderr: %s", stderr)
+		}
```
Comment (farshidtz) [inline utils/exec_test.go:71]: Need to the functions from the testify library to be consistent.

**Review by MonicaisHer — APPROVED** [review]
Thank you, it looks good. I also tested the function for verbose execution, and stdout and stderr are now both shown in the testing logs as follow:
```
utils.ExecVerbose(t, "curl localhost:90)
```
```
exec.go:32: [exec] curl localhost:90
exec.go:118: [stderr]   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
exec.go:118: [stderr]                                  Dload  Upload   Total   Spent    Left  Speed
exec.go:118: [stdout] <!DOCTYPE html>
exec.go:118: [stdout] <!--
exec.go:118: [stdout]   Copyright 2017 Openthread authors. All rights reserved.
```

**Pull request merged**